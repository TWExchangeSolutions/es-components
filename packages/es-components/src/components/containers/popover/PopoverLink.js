import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'lodash';
import styled, { ThemeProvider, withTheme } from 'styled-components';
import uncontrollable from 'uncontrollable';

import defaultTheme from '../../theme/defaultTheme';
import Button from '../../controls/buttons/Button';
import PopoverTrigger from './PopoverTrigger';

const PopoverButton = styled(Button)`
  border-bottom: ${props => (props.suppressUnderline ? 'none' : '1px dotted')};
  text-decoration: none;
`;

export class PopoverLink extends React.Component {
  toggleShow = () => {
    const showPopover = !this.props.showPopover;
    this.props.onToggle(showPopover);
  };

  hidePopover = () => {
    if (this.props.showPopover) {
      this.toggleShow();
    }
  };

  render() {
    const {
      children,
      isLinkButton,
      popoverTitle,
      popoverContent,
      popoverPlacement,
      showPopover,
      showCloseButton,
      suppressUnderline,
      theme,
      buttonStyle,
      ariaLabel
    } = this.props;

    return (
      <ThemeProvider theme={theme}>
        <PopoverTrigger
          onHideOverlay={this.hidePopover}
          popoverContent={popoverContent}
          popoverPlacement={popoverPlacement}
          popoverTarget={this.popoverTarget}
          popoverTitle={popoverTitle}
          showPopover={showPopover}
          showCloseButton={showCloseButton}
        >
          <PopoverButton
            aria-haspopup="dialog"
            data-trigger="focus"
            handleOnClick={this.toggleShow}
            ref={btn => {
              this.popoverTarget = btn;
            }}
            styleType={buttonStyle}
            isLinkButton={isLinkButton}
            suppressUnderline={suppressUnderline}
            aria-label={ariaLabel}
          >
            {children}
          </PopoverButton>
        </PopoverTrigger>
      </ThemeProvider>
    );
  }
}

PopoverLink.propTypes = {
  /** The link content which activates the popover */
  children: PropTypes.node.isRequired,
  /** The text displayed in the popover title section */
  popoverTitle: PropTypes.string,
  /** The content displayed in the popover body */
  popoverContent: PropTypes.node.isRequired,
  /** The placement of the popover in relation to the link */
  popoverPlacement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  /** Function to toggle display of popover */
  onToggle: PropTypes.func,
  /** Displays or hides the popover */
  showPopover: PropTypes.bool,
  /** Display a close ('x') button */
  showCloseButton: PropTypes.bool,
  /** Hide underline from link. Useful for children like Icons */
  suppressUnderline: PropTypes.bool,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object,
  /** The button style of the popover link */
  buttonStyle: PropTypes.string,
  /** Sets the link to use a text rather than a button style **/
  isLinkButton: PropTypes.bool,
  /** Sets the aria-label attribute to allow for textless buttons **/
  ariaLabel: PropTypes.string
};

PopoverLink.defaultProps = {
  popoverPlacement: 'top',
  onToggle: noop,
  showPopover: false,
  showCloseButton: false,
  suppressUnderline: false,
  theme: defaultTheme,
  buttonStyle: 'primary',
  isLinkButton: true
};

const UncontrolledPopoverLink = uncontrollable(PopoverLink, {
  showPopover: 'onToggle'
});

export default withTheme(UncontrolledPopoverLink);
