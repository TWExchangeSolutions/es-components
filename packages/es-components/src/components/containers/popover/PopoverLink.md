```
<p>
  Click
  <PopoverLink
    popoverTitle="Popover"
    popoverContent="This is the popover's content. Anim pariatur cliche reprehenderit, enim eiusmod high life accusamus terry richardson ad squid. 3 wolf moon officia aute, non cupidatat skateboard dolor brunch."
  >
    here
  </PopoverLink> to activate the popover.
</p>
```

```
class PopoverExample extends React.Component {
  constructor() {
    this.state = { show: false };
    this.onPopoverToggled = this.onPopoverToggled.bind(this);
  }

  onPopoverToggled(value) {
    this.setState({ show: value });
    console.log(`Popover ${value}!`);
  };

  render() {
    return (
      <p>
        Example of a
        <PopoverLink
          popoverTitle="Popover"
          popoverContent="Here is some content. Check the console!"
          onToggle={this.onPopoverToggled}
          showPopover={this.state.show}
          showCloseButton
        >
          popover
        </PopoverLink> with an onToggle handler.
      </p>
    )
  }
}
<PopoverExample />
```

```
<p>
  Click the question mark for some help text.
  <PopoverLink popoverContent="Here's some help text" suppressUnderline><Icon name="question-sign" /></PopoverLink>
</p>
```
