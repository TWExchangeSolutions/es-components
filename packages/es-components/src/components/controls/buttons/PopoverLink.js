import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import classnames from 'classnames';
import LinkButton from './LinkButton';
import { useTheme } from '../../util/useTheme';

const StyledButton = styled(LinkButton)`
  border-bottom-width: 1px;
  border-bottom-style: dashed;
  border-bottom-color: ${props =>
    props.suppressUnderline ? 'transparent' : 'initial'};
  margin-bottom: 2px;
  text-decoration: none;

  &:hover {
    border-bottom-style: solid;
  }
`;

function InnerButton({
  children,
  styleType,
  innerRef,
  suppressUnderline,
  ...buttonProps
}) {
  const theme = useTheme();
  const variant = theme.buttonStyles.button.variant[styleType];
  const { className, ...otherProps } = buttonProps;
  const sharedProps = {
    variant,
    suppressUnderline,
    ref: innerRef,
    ...otherProps
  };

  return (
    <StyledButton
      {...sharedProps}
      className={classnames('es-button es-button--popover-trigger', className)}
    >
      {children}
    </StyledButton>
  );
}

const PopoverLink = React.forwardRef((props, ref) => (
  <InnerButton innerRef={ref} {...props} />
));

PopoverLink.propTypes = {
  children: PropTypes.node.isRequired,
  /** Select the color style of the button, types come from theme */
  styleType: PropTypes.string,
  /** Hide underline from link. Useful for children like Icons */
  suppressUnderline: PropTypes.bool
};

PopoverLink.defaultProps = {
  styleType: 'default',
  block: false,
  size: 'default',
  suppressUnderline: false
};

export default PopoverLink;
