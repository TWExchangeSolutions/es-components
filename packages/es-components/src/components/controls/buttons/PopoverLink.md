A specially styled `LinkButton` that can be used with `Popover` if a text link is desired over the normal `Button`.

```
<PopoverLink styleType="success">Popover Link Button</PopoverLink>
```

You can optionally suppress the underline in the link.

```
<PopoverLink suppressUnderline styleType="success">Popover Link Button</PopoverLink>
```
