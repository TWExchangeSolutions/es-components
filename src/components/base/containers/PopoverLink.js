import React from 'react';
import PropTypes from 'prop-types';
import Overlay from 'react-overlays/lib/Overlay';
import { noop } from 'lodash';
import Button from '../../controls/buttons/Button';
import Popover from './Popover';
import FocusTrap from 'focus-trap-react';
import Fade from '../../util/Fade';
import { injectGlobal } from 'styled-components';

/* eslint-disable no-unused-expressions */
injectGlobal`
  .popover-transition-out {
    opacity: 0;
    transition: opacity 200ms linear;
  }

  .popover-transition-in {
    opacity: 1;
  }
`;
/* eslint-enable no-unused-expressions */

function getArrowPlacement(popoverPlacement) {
  switch (popoverPlacement) {
    case 'top':
    case 'bottom':
      return popoverPlacement;
    case 'right':
      return 'left';
    case 'left':
      return 'right';
    default:
      return popoverPlacement;
  }
}

/*
  monkey-patching the FocusTrap component
  fixes a timing issue with the popover
*/
FocusTrap.prototype.componentDidMount = function componentDidMount() {
  const specifiedFocusTrapOptions = this.props.focusTrapOptions;
  const tailoredFocusTrapOptions = {
    returnFocusOnDeactivate: false
  };
  for (const optionName in specifiedFocusTrapOptions) {
    if (!specifiedFocusTrapOptions.hasOwnProperty(optionName)) continue;
    if (optionName === 'returnFocusOnDeactivate') continue;
    tailoredFocusTrapOptions[optionName] =
      specifiedFocusTrapOptions[optionName];
  }
  this.focusTrap = this.props._createFocusTrap(
    this.node,
    tailoredFocusTrapOptions
  );
  if (this.props.active) {
    setTimeout(() => {
      this.focusTrap.activate();
    }, 0);
  }
  if (this.props.paused) {
    this.focusTrap.pause();
  }
};

const FadeTransition = props => (
  <Fade
    transitionClassOut="popover-transition-out"
    transitionClassIn="popover-transition-in"
    {...props}
  />
);

class PopoverLink extends React.Component {
  constructor(props) {
    super(props);

    this.state = { isOpen: false };
  }

  togglePopover = () => {
    if (this.state.isOpen) {
      this.hidePopover();
    } else {
      this.setState({ isOpen: true });
    }
  };

  hidePopover = () => {
    /*
      The focus trap calls this twice for some reason, even if the onHide call
      is removed from the Overlay. Checking so onPopoverHidden isn't called twice.
    */
    if (this.state.isOpen) {
      this.setState({ isOpen: false });
      this.props.onPopoverHidden();
    }
  };

  render() {
    const {
      children,
      popoverTitle,
      popoverContent,
      popoverPlacement,
      containsFormElement
    } = this.props;

    const { isOpen } = this.state;
    const arrowPlacement = getArrowPlacement(popoverPlacement);
    const focusTrapOptions = {
      onDeactivate: this.hidePopover
    };

    return (
      <span>
        <span
          ref={span => {
            this.popoverTarget = span;
          }}
        >
          <Button
            data-trigger="focus"
            styleType="link"
            handleOnClick={this.togglePopover}
          >
            {children}
          </Button>
        </span>

        <Overlay
          show={isOpen}
          placement={popoverPlacement}
          target={this.popoverTarget}
          rootClose={!containsFormElement}
          onHide={this.hidePopover}
          transition={FadeTransition}
        >
          <Popover
            title={popoverTitle}
            arrowPlacement={arrowPlacement}
            containsFormElement={containsFormElement}
          >
            <FocusTrap
              active={containsFormElement}
              focusTrapOptions={focusTrapOptions}
            >
              {popoverContent}
            </FocusTrap>
          </Popover>
        </Overlay>
      </span>
    );
  }
}

PopoverLink.propTypes = {
  /** The link content which activates the popover */
  children: PropTypes.node.isRequired,
  /** The text displayed in the popover title section */
  popoverTitle: PropTypes.string,
  /** The content displayed in the popover body */
  popoverContent: PropTypes.node.isRequired,
  /** The placement of the popover in relation to the link */
  popoverPlacement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  containsFormElement: PropTypes.bool,
  /** Function to run when the popover is hidden */
  onPopoverHidden: PropTypes.func
};

PopoverLink.defaultProps = {
  popoverPlacement: 'top',
  containsFormElement: false,
  onPopoverHidden: noop
};

export default PopoverLink;
