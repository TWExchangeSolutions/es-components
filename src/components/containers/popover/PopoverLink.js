import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'lodash';
import styled from 'styled-components';
import uncontrollable from 'uncontrollable';

import Button from '../../controls/buttons/Button';
import PopoverTrigger from './PopoverTrigger';

const PopoverButton = styled(Button)`
  border-bottom: ${props => (props.suppressUnderline ? 'none' : '1px dotted')};
  text-decoration: none;
`;

export class PopoverLink extends React.Component {
  toggleShow = () => {
    const showPopover = !this.props.showPopover;

    this.props.onToggle(showPopover);

    if (!showPopover) {
      this.props.onPopoverHidden();
    }
  };

  hidePopover = () => {
    /*
      The focus trap calls this twice for some reason, even if the onHide call
      is removed from the Overlay. Checking so onPopoverHidden isn't called twice.
    */
    if (this.props.showPopover) {
      this.toggleShow();
    }
  };

  render() {
    const {
      children,
      popoverTitle,
      popoverContent,
      popoverPlacement,
      showPopover,
      suppressCloseButton,
      suppressUnderline
    } = this.props;

    return (
      <PopoverTrigger
        onHideOverlay={this.hidePopover}
        popoverContent={popoverContent}
        popoverPlacement={popoverPlacement}
        popoverTarget={this.popoverTarget}
        popoverTitle={popoverTitle}
        showPopover={showPopover}
        suppressCloseButton={suppressCloseButton}
      >
        <PopoverButton
          aria-haspopup="dialog"
          data-trigger="focus"
          handleOnClick={this.toggleShow}
          ref={btn => {
            this.popoverTarget = btn;
          }}
          styleType="link"
          suppressUnderline={suppressUnderline}
        >
          {children}
        </PopoverButton>
      </PopoverTrigger>
    );
  }
}

PopoverLink.propTypes = {
  /** The link content which activates the popover */
  children: PropTypes.node.isRequired,
  /** The text displayed in the popover title section */
  popoverTitle: PropTypes.string,
  /** The content displayed in the popover body */
  popoverContent: PropTypes.node.isRequired,
  /** The placement of the popover in relation to the link */
  popoverPlacement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  /** Function called when the popover is hidden */
  onPopoverHidden: PropTypes.func,
  /** Function to toggle display of popover */
  onToggle: PropTypes.func,
  /** Displays or hides the popover */
  showPopover: PropTypes.bool,
  /** Hide the default 'x' close button */
  suppressCloseButton: PropTypes.bool,
  /** Hide underline from link. Useful for children like Icons */
  suppressUnderline: PropTypes.bool
};

PopoverLink.defaultProps = {
  popoverPlacement: 'top',
  onPopoverHidden: noop,
  onToggle: noop,
  showPopover: false,
  suppressCloseButton: true,
  suppressUnderline: false
};

const UncontrolledPopoverLink = uncontrollable(PopoverLink, {
  showPopover: 'onToggle'
});

export default UncontrolledPopoverLink;
