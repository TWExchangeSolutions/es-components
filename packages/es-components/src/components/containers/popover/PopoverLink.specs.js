/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import PopoverLink from './PopoverLink';

describe('popoverLink component', () => {
  const onPopoverHidden = jest.fn();
  const onToggle = jest.fn();

  const instanceToRender = (
    <PopoverLink
      popoverTitle="Popover Title"
      popoverContent="This is a popover."
      onPopoverHidden={onPopoverHidden}
      onToggle={onToggle}
    >
      This is the popover link text.
    </PopoverLink>
  );

  it('renders as expected', () => {
    const tree = renderer.create(instanceToRender).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('calls onToggle when link is clicked', () => {
    const popoverInstance = mount(instanceToRender);
    const button = popoverInstance.find('[data-trigger="focus"]').hostNodes();

    button.simulate('click');
    expect(onToggle).toHaveBeenCalled();
  });
});
