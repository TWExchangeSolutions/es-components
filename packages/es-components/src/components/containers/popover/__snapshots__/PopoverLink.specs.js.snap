// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`popoverLink component renders as expected 1`] = `
<span>
  <button
    aria-haspopup="dialog"
    aria-label={undefined}
    className="sc-jzJRlG gUzwcG sc-bxivhb eboPuZ sc-bdVaJa jXnGOW"
    data-trigger="focus"
    onClick={[Function]}
  >
    This is the popover link text.
  </button>
</span>
`;
